cmd/ceremony/cert.go

```diff
@@ -333,3 +366,128 @@ func generateCSR(profile *certProfile, signer crypto.Signer) ([]byte, error) {
 	}
 	return csrDER, nil
 }
+
+// issueLintCert issues a linting certificate from a given template certificate
+// signed by a given issuer and returns the certificate or an error. The public
+// key from the loaded certificate is checked by the GoodKey package.
+func issueLintCert(tbs, issuer *x509.Certificate, subjectPubKey crypto.PublicKey, signer crypto.Signer, skipLints []string) (*x509.Certificate, error) {
+	lintCertBytes, err := linter.Check(tbs, subjectPubKey, issuer, signer, skipLints)
+	if err != nil {
+		return nil, fmt.Errorf("certificate failed pre-issuance lint: %w", err)
+	}
+
+	lintCert, err := x509.ParseCertificate(lintCertBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	err = kp.GoodKey(context.Background(), lintCert.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return lintCert, nil
+}
+
+func equalPubKeys(a, b interface{}) bool {
+	aBytes, err := x509.MarshalPKIXPublicKey(a)
+	if err != nil {
+		return false
+	}
+	bBytes, err := x509.MarshalPKIXPublicKey(b)
+	if err != nil {
+		return false
+	}
+	return bytes.Equal(aBytes, bBytes)
+}
+
+func openSigner(cfg PKCS11SigningConfig, pubKey crypto.PublicKey) (crypto.Signer, *hsmRandReader, error) {
+	session, err := pkcs11helpers.Initialize(cfg.Module, cfg.SigningSlot, cfg.PIN)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to setup session and PKCS#11 context for slot %d: %s",
+			cfg.SigningSlot, err)
+	}
+	log.Printf("Opened PKCS#11 session for slot %d\n", cfg.SigningSlot)
+	signer, err := session.NewSigner(cfg.SigningLabel, pubKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to retrieve private key handle: %s", err)
+	}
+	if !equalPubKeys(signer.Public(), pubKey) {
+		return nil, nil, fmt.Errorf("signer pubkey did not match issuer pubkey")
+	}
+	log.Println("Retrieved private key handle")
+	return signer, newRandReader(session), nil
+}
+
+// loadCert loads a PEM certificate specified by filename or returns an error.
+// The public key from the loaded certificate is checked by the GoodKey package.
+func loadCert(filename string) (*x509.Certificate, error) {
+	certPEM, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("Loaded certificate from %s\n", filename)
+	block, _ := pem.Decode(certPEM)
+	if block == nil {
+		return nil, fmt.Errorf("No data in cert PEM file %s", filename)
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	goodkeyErr := kp.GoodKey(context.Background(), cert.PublicKey)
+	if goodkeyErr != nil {
+		return nil, goodkeyErr
+	}
+
+	return cert, nil
+}
+
+func signAndWriteCert(tbs, issuer *x509.Certificate, subjectPubKey crypto.PublicKey, signer crypto.Signer, certPath string) (*x509.Certificate, error) {
```

Comment:
Sure, can do. My eyes were crossing why jumping between all the helper functions and the cross-certificate that this was my _1_ solution instead of something reasonable like opening a new tab.